keywrap/pkcs11/keywrapper_pkcs11.go

```diff
@@ -0,0 +1,130 @@
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package pkcs11
+
+import (
+	"github.com/containers/ocicrypt/config"
+	"github.com/containers/ocicrypt/crypto/pkcs11"
+	"github.com/containers/ocicrypt/keywrap"
+	"github.com/containers/ocicrypt/utils"
+
+	"github.com/pkg/errors"
+	pkcs11uri "github.com/stefanberger/go-pkcs11uri"
+)
+
+type pkcs11KeyWrapper struct {
+}
+
+func (kw *pkcs11KeyWrapper) GetAnnotationID() string {
+	return "org.opencontainers.image.enc.keys.experimental.pkcs11"
+}
+
+// NewKeyWrapper returns a new key wrapping interface using pkcs11
+func NewKeyWrapper() keywrap.KeyWrapper {
+	return &pkcs11KeyWrapper{}
+}
+
+// WrapKeys wraps the session key for recpients and encrypts the optsData, which
+// describe the symmetric key used for encrypting the layer
+func (kw *pkcs11KeyWrapper) WrapKeys(ec *config.EncryptConfig, optsData []byte) ([]byte, error) {
+	// FIXME: The pkcs11-uris may be insufficient for more advanced pkcs11 modules that need
+	// environment variables set. So, public keys may have to be 'pkcs11-yamls' as well, so
+	// that we can set environment variables for the pkcs11 module to find config file etc.
+	pkcs11Recipients, err := addPubKeys(&ec.DecryptConfig, append(ec.Parameters["pkcs11-pubkeys"], ec.Parameters["pkcs11-uris"]...))
+	if err != nil {
+		return nil, err
+	}
+	// no recipients is not an error...
+	if len(pkcs11Recipients) == 0 {
+		return nil, nil
+	}
+
+	jsonString, err := pkcs11.EncryptMultiple(pkcs11Recipients, optsData)
+	if err != nil {
+		return nil, errors.Wrapf(err, "PKCS11 EncryptMulitple failed")
+	}
+	return jsonString, nil
+}
+
+func (kw *pkcs11KeyWrapper) UnwrapKey(dc *config.DecryptConfig, jsonString []byte) ([]byte, error) {
+	var pkcs11PrivKeys []*pkcs11uri.Pkcs11URI
+
+	privKeys := kw.GetPrivateKeys(dc.Parameters)
+	if len(privKeys) == 0 {
+		return nil, errors.New("No private keys found for PKCS11 decryption")
+	}
+	for _, privKey := range privKeys {
+		key, err := utils.ParsePrivateKey(privKey, nil, "PKCS11")
+		if err != nil {
+			return nil, err
+		}
+		switch pkcs11Key := key.(type) {
+		case *pkcs11.Pkcs11KeyFileObject:
+			if dc.Pkcs11Config != nil {
+				pkcs11Key.Uri.SetModuleDirectories(dc.Pkcs11Config.ModuleDirectories)
+			}
+			pkcs11PrivKeys = append(pkcs11PrivKeys, pkcs11Key.Uri)
+		default:
+			continue
+		}
+	}
+
+	plaintext, err := pkcs11.Decrypt(pkcs11PrivKeys, jsonString)
+	if err == nil {
+		return plaintext, nil
+	}
+
+	return nil, errors.New("PKCS11: No suitable private key found for decryption")
+}
+
+func (kw *pkcs11KeyWrapper) NoPossibleKeys(dcparameters map[string][][]byte) bool {
+	return len(kw.GetPrivateKeys(dcparameters)) == 0
+}
+
+func (kw *pkcs11KeyWrapper) GetPrivateKeys(dcparameters map[string][][]byte) [][]byte {
+	return dcparameters["pkcs11-yamls"]
```

Comment:
I think this interface is meant to actually return the bytes of private keys that can be parsed in PEM/DER format. It is fine for this method to not be implemented as it is optional though.

Reply:
No, this is a pkcs11 YAML file representing a private key.

Reply:
lets follow the optional implementation then

Reply:
What is 'optional implementation' ?

Reply:
Leave as is so others can get the raw bytes of the pkcs11 private key YAML file and parse it if necessary?

Reply:
I'm fine with this, I think we can update https://github.com/containers/ocicrypt/blob/3c3f2f1f1820aded1c173c9d6173c85cd0158935/keywrap/keywrap.go#L34

to be more specific that it should not expect the private key returned to be a standard DER/PEM format and should only used for convenience specific to each protocol.. that way we don't end up with people using this public interface to actually parse DER/PEM hoping to do something useful.